src/daft-core/src/array/ops/any_value.rs

```diff
@@ -0,0 +1,104 @@
+use common_error::DaftResult;
+
+use crate::{
+    array::{DataArray, FixedSizeListArray, ListArray, StructArray},
+    datatypes::*,
+};
+
+use super::{full::FullNull, DaftAnyValueAggable, GroupIndices};
+
+fn get_any_grouped_idx(
+    groups: &GroupIndices,
+    ignore_nulls: bool,
+    validity: Option<&Bitmap>,
+) -> UInt64Array {
+    let group_indices = if ignore_nulls && let Some(validity) = validity {
+        Box::new(PrimitiveArray::from_trusted_len_iter(groups.iter().map(|g| {
+            for i in g {
+                if validity.get_bit(*i as usize) {
+                    return Some(*i)
+                }
+            }
+            None
+        })))
+    } else {
+        Box::new(PrimitiveArray::from_trusted_len_iter(groups.iter().map(|g| g.first().cloned())))
+    };
+
+    DataArray::from(("", group_indices))
+}
+
+impl<T> DaftAnyValueAggable for DataArray<T>
+where
+    T: DaftNumericType,
+{
+    type Output = DaftResult<DataArray<T>>;
+
+    fn any_value(&self, ignore_nulls: bool) -> Self::Output {
+        if ignore_nulls && let Some(validity) = self.validity() {
+            for i in 0..self.len() {
+                if validity.get_bit(i) {
+                    return self.slice(i, i+1);
+                }
+            }
+
+            Ok(DataArray::full_null(
+                self.name(),
+                self.data_type(),
+                1
+            ))
+        } else {
+            self.slice(0, 1)
+        }
+    }
+
+    fn grouped_any_value(&self, groups: &GroupIndices, ignore_nulls: bool) -> Self::Output {
+        self.take(&get_any_grouped_idx(groups, ignore_nulls, self.validity()))
+    }
+}
+
+macro_rules! impl_daft_any_value {
+    ($arr:ident) => {
+
+        // there is a strange bug in the formatter which adds extra indentations to any_value
+        #[rustfmt::skip]
+        impl DaftAnyValueAggable for $arr {
+            type Output = DaftResult<$arr>;
+
+            fn any_value(&self, ignore_nulls: bool) -> Self::Output {
+                if ignore_nulls && let Some(validity) = self.validity() {
+                    for i in 0..self.len() {
+                        if validity.get_bit(i) {
+                            return self.slice(i, i+1);
+                        }
+                    }
+
+                    Ok($arr::full_null(
+                        self.name(),
+                        self.data_type(),
+                        1
+                    ))
+                } else {
+                    self.slice(0, 1)
+                }
+            }
+
+            fn grouped_any_value(&self, groups: &GroupIndices, ignore_nulls: bool) -> Self::Output {
+                self.take(&get_any_grouped_idx(groups, ignore_nulls, self.validity()))
+            }
+        }
+    };
+}
+
+impl_daft_any_value!(Utf8Array);
+impl_daft_any_value!(BooleanArray);
+impl_daft_any_value!(BinaryArray);
+impl_daft_any_value!(NullArray);
+impl_daft_any_value!(ExtensionArray);
```

Comment:
@samster25 I recall you suggested implementing this for some generic trait instead of using macros, but when I tried to implement this for `DataArray<T> where T: DaftPhysicalType`, calling `self.take` gave me this compiler error: 

> the method `take` exists for reference `&DataArray<T>`, but its trait bounds were not satisfied
the following trait bounds were not satisfied:
`&DataArray<T>: Iterator`
which is required by `&mut &DataArray<T>: Iterator`
`DataArray<T>: Iterator`
which is required by `&mut DataArray<T>: Iterator`

I'm not sure what to make of it, my guess is that since `take` was implemented independently for `DataArray<DaftNumericType>` and the other DataArray types instead of some more generic trait, the compiler isn't willing to claim that every `DataArray<DaftPhysicalType>` implements this.

Reply:
Ah yes, that's due to rust being unable to solve the trait bounds `DataArray<DaftPhysicalType>` since we don't have a unified `impl<T:DaftPhysicalType> Take for DataArray<T>`